#Python Task 1 – "Customer Complaint Analyzer"
Scenario:
You work at a company that receives daily customer complaints. You’re given a list of complaint texts, and your task is to analyze which keyword appears most often (to identify common issues).

Input:
A list of customer complaints like:

In [11]:
complaints = [
    "App crashes after login",
    "Login fails on desktop version",
    "App crashes randomly during use",
    "Slow loading on app",
    "Login issues again",
    "Crash happens after I log in",
    "Loading takes forever"
]

# Task
processes text data to find the top 3 most frequent keywords, ignoring common English stopwords. The code should lowercase the text, remove punctuation, tokenize words, and use `collections.Counter` for frequency counting.

## Define stopwords

### Subtask:
Create a list of common English stopwords to be ignored.


**Reasoning**:
Create a list of common English stopwords as instructed.



In [12]:
import nltk
from nltk.corpus import stopwords

nltk.download('stopwords')
print(stopwords.words('english'))

['a', 'about', 'above', 'after', 'again', 'against', 'ain', 'all', 'am', 'an', 'and', 'any', 'are', 'aren', "aren't", 'as', 'at', 'be', 'because', 'been', 'before', 'being', 'below', 'between', 'both', 'but', 'by', 'can', 'couldn', "couldn't", 'd', 'did', 'didn', "didn't", 'do', 'does', 'doesn', "doesn't", 'doing', 'don', "don't", 'down', 'during', 'each', 'few', 'for', 'from', 'further', 'had', 'hadn', "hadn't", 'has', 'hasn', "hasn't", 'have', 'haven', "haven't", 'having', 'he', "he'd", "he'll", 'her', 'here', 'hers', 'herself', "he's", 'him', 'himself', 'his', 'how', 'i', "i'd", 'if', "i'll", "i'm", 'in', 'into', 'is', 'isn', "isn't", 'it', "it'd", "it'll", "it's", 'its', 'itself', "i've", 'just', 'll', 'm', 'ma', 'me', 'mightn', "mightn't", 'more', 'most', 'mustn', "mustn't", 'my', 'myself', 'needn', "needn't", 'no', 'nor', 'not', 'now', 'o', 'of', 'off', 'on', 'once', 'only', 'or', 'other', 'our', 'ours', 'ourselves', 'out', 'over', 'own', 're', 's', 'same', 'shan', "shan't", 'she

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [13]:
stopwords = [
    'i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd",
    'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers',
    'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which',
    'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been',
    'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if',
    'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between',
    'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out',
    'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why',
    'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not',
    'only', 'own', 'same', 'so', 'than', 'too', 'very', 's', 't', 'can', 'will', 'just', 'don', "don't", 'should',
    "should've", 'now', 'd', 'll', 'm', 'o', 're', 've', 'y', 'ain', 'aren', "aren't", 'couldn', "couldn't",
    'didn', "didn't", 'doesn', "doesn't", 'hadn', "hadn't", 'hasn', "hasn't", 'haven', "haven't", 'isn', "isn't",
    'ma', 'mightn', "mightn't", 'mustn', "mustn't", 'needn', "needn't", 'shan', "shan't", 'shouldn', "shouldn't",
    'wasn', "wasn't", 'weren', "weren't", 'won', "won't", 'wouldn', "wouldn't"
]
print(f"Created a list of {len(stopwords)} stopwords.")

Created a list of 179 stopwords.


## Clean and tokenize

### Subtask:
Process each complaint string by converting to lowercase, removing punctuation, and splitting into individual words (tokens).


**Reasoning**:
Process each complaint by converting to lowercase, removing punctuation, and splitting into tokens.



In [14]:
import string
all_tokens = []

for complaint in complaints:
  # convert to lower case
  lower_complaint = complaint.lower()
  #remove punctualtion
  cleaned_complaint = '' .join(char for char in lower_complaint if char not in string.punctuation)
  # split into tokens
  tokens = cleaned_complaint.split()
  #Extend the all_token list
  all_tokens.extend(tokens)
print(all_tokens[:20])

['app', 'crashes', 'after', 'login', 'login', 'fails', 'on', 'desktop', 'version', 'app', 'crashes', 'randomly', 'during', 'use', 'slow', 'loading', 'on', 'app', 'login', 'issues']


## Filter stopwords

### Subtask:
Remove the defined stopwords from the list of tokens.


**Reasoning**:
Iterate through the list of tokens and remove stopwords.



In [4]:
filtered_tokens = [token for token in all_tokens if token not in stopwords]
print(filtered_tokens[:20])

['app', 'crashes', 'login', 'login', 'fails', 'desktop', 'version', 'app', 'crashes', 'randomly', 'use', 'slow', 'loading', 'app', 'login', 'issues', 'crash', 'happens', 'log', 'loading']


## Count word frequency

### Subtask:
Use `collections.Counter` to count the occurrences of each word.


**Reasoning**:
Import the Counter class, create a Counter object from the filtered tokens, and print the word counts to complete the subtask of counting word occurrences.



In [15]:
from collections import Counter

word_counts = Counter(filtered_tokens)
print(word_counts)

Counter({'app': 3, 'login': 3, 'crashes': 2, 'loading': 2, 'fails': 1, 'desktop': 1, 'version': 1, 'randomly': 1, 'use': 1, 'slow': 1, 'issues': 1, 'crash': 1, 'happens': 1, 'log': 1, 'takes': 1, 'forever': 1})


## Print top keywords

### Subtask:
Identify and print the top 3 most frequent words.


**Reasoning**:
Use the `most_common()` method of the `word_counts` Counter object to find and print the top 3 most frequent words.



In [6]:
top_3_words = word_counts.most_common(3)
print(top_3_words)

[('app', 3), ('login', 3), ('crashes', 2)]


## Summary:

### Data Analysis Key Findings

*   A list of 179 common English stopwords was defined.
*   The text data was successfully cleaned by converting to lowercase and removing punctuation.
*   The cleaned text was tokenized into individual words.
*   Stopwords were filtered out from the list of tokens.
*   Word frequencies were counted using `collections.Counter`.
*   The top 3 most frequent words found were 'app' (3 occurrences), 'login' (3 occurrences), and 'crashes' (2 occurrences).

---------------------

#Python Task 2 – "Product Sales Aggregator"
Scenario:
You’re working with an e-commerce platform. Each day, they collect logs of orders. Your job is to aggregate the total quantity sold per product, and sort the products by most sold.

Input:
You get a list of transactions like this:

In [16]:
orders = [
    {"product": "Laptop", "quantity": 2},
    {"product": "Phone", "quantity": 5},
    {"product": "Laptop", "quantity": 1},
    {"product": "Tablet", "quantity": 3},
    {"product": "Phone", "quantity": 2},
    {"product": "Tablet", "quantity": 2},
    {"product": "Phone", "quantity": 1}
]
